ring/ingester_lifecycle.go

```diff
@@ -16,125 +16,212 @@
 package ring
 
 import (
-	"encoding/json"
 	"fmt"
-	"hash/fnv"
 	"math/rand"
 	"net"
 	"os"
+	"sort"
 	"sync"
 	"time"
 
 	"github.com/prometheus/common/log"
 )
 
 const (
-	infName = "eth0"
+	infName           = "eth0"
+	ring              = "ring"
+	heartbeatInterval = 5 * time.Second
 )
 
 // IngesterRegistration manages the connection between the ingester and Consul.
 type IngesterRegistration struct {
-	consul ConsulClient
-	id     string
-	desc   []byte
+	consul    ConsulClient
+	numTokens int
 
-	quit chan struct{}
-	wait sync.WaitGroup
+	id       string
+	hostname string
+	quit     chan struct{}
+	wait     sync.WaitGroup
 }
 
 // RegisterIngester registers an ingester with Consul.
 func RegisterIngester(consulClient ConsulClient, listenPort, numTokens int) (*IngesterRegistration, error) {
-	desc, err := describeLocalIngester(listenPort, numTokens)
+	hostname, err := os.Hostname()
 	if err != nil {
 		return nil, err
 	}
-	buf, err := json.Marshal(desc)
+
+	addr, err := getFirstAddressOf(infName)
 	if err != nil {
 		return nil, err
 	}
 
 	r := &IngesterRegistration{
-		consul: consulClient,
-		id:     desc.ID,
-		desc:   buf,
-		quit:   make(chan struct{}),
+		consul:    consulClient,
+		numTokens: numTokens,
+
+		id:       hostname,
+		hostname: fmt.Sprintf("%s:%d", addr, listenPort),
+		quit:     make(chan struct{}),
 	}
+
 	r.wait.Add(1)
-	go r.updateLoop()
+	go r.loop()
 	return r, nil
 }
 
-func (r *IngesterRegistration) updateLoop() {
+func (r *IngesterRegistration) loop() {
 	defer r.wait.Done()
-	ticker := time.NewTicker(1 * time.Second)
+	tokens := r.pickTokens()
+	defer r.unregister(tokens)
+	r.heartbeat(tokens)
+}
+
+func (r *IngesterRegistration) pickTokens() []uint32 {
+	var tokens []uint32
+	if err := r.consul.CAS(ring, descFactory, func(in interface{}) (out interface{}, retry bool, err error) {
+		var ringDesc *Desc
+		if in == nil {
+			ringDesc = newDesc()
+		} else {
+			ringDesc = in.(*Desc)
+		}
+
+		takenTokens := []uint32{}
+		for _, token := range ringDesc.Tokens {
+			takenTokens = append(takenTokens, token.Token)
+		}
+		tokens = generateTokens(r.numTokens, takenTokens)
+
+		populateRingDesc(ringDesc, r.id, r.hostname, tokens)
+
+		return ringDesc, true, nil
+	}); err != nil {
```

Comment:
This took me quite a while to figure out. Here are some things that I think might help make it easier to follow:

* Separate assigning to `err` from the `err != nil` check
* Define the function being passed to `CAS` separately and give it a name. Make `tokens` a parameter of that function and close over `tokens` here.